### PR TITLE
feature(whl): add code demo for lm-env

### PR DESCRIPTION
**Pull request by kxzxvbk**

### CODE REVIEW

File: chapter7_tricks/lm_env.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+PyTorch implementation of Language Model Environment.
+
+There are two main components in this documentation:
+- We use GPT-2 as the base language model and construct an environment.
+- We make a demonstration of this environment and users can type prompts in the command line to interact with the language model.
+"""
+import torch
+import gym
+from typing import Callable, Optional, Dict, Tuple
+from transformers import GPT2Tokenizer, GPT2LMHeadModel
```
Comment (nighood) [inline chapter7_tricks/lm_env.py:11]: When I looked at this code example, I was mainly not clear about the use of GPT2 related classes (sorry for some problems using huddingface). For example, what does the output of the "transformers.GPT2LMHeadModel.generate" method look like. So maybe here can be added a link like this: https://huggingface.co/transformers/v3.0.2/model_doc/gpt2.html#gpt2lmheadmodel

File: chapter7_tricks/lm_env.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+class TextHistory:
+    """
+    **Overview:**
+        The TextHistory class keeps track of the history of an interaction between the language model and the environment.
+    """
+
+    def __init__(self, text: str, tokens: torch.Tensor):
+        """
+        Initialize TextHistory.
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:23]: overview

File: chapter7_tricks/lm_env.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        The TextEnvironment enables interaction of a LLM with an environment.
+    """
+
+    def __init__(self, model: GPT2LMHeadModel, tokenizer: GPT2Tokenizer, reward_fn: Callable,
+                 max_turns: int = 4, generation_kwargs: Optional[Dict] = None):
+        """
+        **Overview:**
+            Initialize the TextEnvironment.
+
+        Arguments:
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:84]: also use **

File: chapter7_tricks/lm_env.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        rew = self.reward_fn(self.history.last_text_segment)
+        self.turn += 1
+        self.history.completed = self.turn >= self.max_turns
+        return self.history, rew, self.history.completed
+
+    # delimiter
+    def _generate(self, query_tensors: torch.Tensor) -> torch.Tensor:
+        """
+        Generate responses for a list of query tensors.
+        args:
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:158]: Arguments

File: chapter7_tricks/lm_env.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        self.generate()
+        rew = self.reward_fn(self.history.last_text_segment)
+        self.turn += 1
+        self.history.completed = self.turn >= self.max_turns
+        return self.history, rew, self.history.completed
+
+    # delimiter
+    def _generate(self, query_tensors: torch.Tensor) -> torch.Tensor:
+        """
+        Generate responses for a list of query tensors.
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:157]: Overview

File: chapter7_tricks/lm_env.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        args:
+            query_tensors (list[torch.Tensor]): A list of query tensors to generate responses for.
+        """
+        query_tensors = query_tensors.unsqueeze(0)
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:161]: add more comments

File: chapter7_tricks/lm_env.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        # The env step once.
+        his, reward, done = env.step(q)
+        # Extract and print the response.
+        print(f"Response: {his.last_text_segment}")
+        # If the environment is done, break the interaction loop.
+        if done:
+            break
+
+
+if __name__ == '__main__':
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:214]: provide some input examples

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: chapter7_tricks/lm_env.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            self.generation_kwargs = dict()
+        else:
+            self.generation_kwargs = generation_kwargs
+        
+        # Count the times of ``self.step()``
+        self.turn = 0
+        # Preserve the history of interactions.
+        self.history = None
+        # Determine the device of running the model (cpu or cuda).
+        self.current_device = self.model.device
```
Comment (PaParaZz1) [inline chapter7_tricks/lm_env.py:108]: define observartion space, action space and reward space

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]
